# 1 

We can interpret this as people spend on average $873 if they make no money and have no children recreationally. as their income increases, so does the amount that they spend. this makes sense because they have more disposable income. The more children that a family has, the less money that a family will spend. This is due to them having to spend more money on the child and saving up for future expenses such as college

we need the p-value to determine if these features are statistically significant

In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
df = pd.read_sql_query('select * from weatherinszeged',con=engine)

engine.dispose()

In [6]:
X = df[['humidity', 'windspeed']]
Y = df['apparenttemperature'] - df['temperature']
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 23 Feb 2020   Prob (F-statistic):               0.00
Time:                        21:12:28   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Yes both coefficients are statistically significant with a p-value of 0. When the target variable is positive, it means that the perceived temperature is hotter than it actually is. Also when the target variable is negative, then the perceived temperature is colder than it actually is. Both coefficients are negative. This means that as humidity increases, people think that it's colder than it is. Also when windspeed increases, people think it is colder than it is. the wind speed makes sense as more wind increases wind chill. the humidity can cause people to think that the temperature is colder or hotter depending on the temperature. a higher humidity makes people think it is colder than a higher windspeed, according to the coefficients. 

In [14]:
df['humid_wind'] = df.windspeed * df.humidity

X = df[['humidity', 'windspeed', 'humid_wind']]
Y = df['apparenttemperature'] - df['temperature']
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 23 Feb 2020   Prob (F-statistic):               0.00
Time:                        21:25:41   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0839      0.033      2.511      0.012       0.018       0.149
humidity       0.1775      0.043      4.133      0.000       0.093       0.262
windspeed      0.0905      0.002     36.797      0.000       0.086       0.095
humid_wind    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

the coefficients are all statistically significant. the signs of humidity and windspeed are positive while humidity_wind interaction is negative. the humidity_wind interaction is the largest and shows that it affects the target variable the most. 

In [15]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
df = pd.read_sql_query('select * from houseprices',con=engine)

engine.dispose()

In [16]:
numeric_columns = df.select_dtypes(['int64', 'float64']).columns
np.abs(df[numeric_columns].iloc[:,1:].corr().loc[:,"saleprice"]).sort_values(ascending=False)

saleprice        1.000000
overallqual      0.790982
grlivarea        0.708624
garagecars       0.640409
garagearea       0.623431
totalbsmtsf      0.613581
firstflrsf       0.605852
fullbath         0.560664
totrmsabvgrd     0.533723
yearbuilt        0.522897
yearremodadd     0.507101
garageyrblt      0.486362
masvnrarea       0.477493
fireplaces       0.466929
bsmtfinsf1       0.386420
lotfrontage      0.351799
wooddecksf       0.324413
secondflrsf      0.319334
openporchsf      0.315856
halfbath         0.284108
lotarea          0.263843
bsmtfullbath     0.227122
bsmtunfsf        0.214479
bedroomabvgr     0.168213
kitchenabvgr     0.135907
enclosedporch    0.128578
screenporch      0.111447
poolarea         0.092404
mssubclass       0.084284
overallcond      0.077856
mosold           0.046432
threessnporch    0.044584
yrsold           0.028923
lowqualfinsf     0.025606
miscval          0.021190
bsmthalfbath     0.016844
bsmtfinsf2       0.011378
Name: saleprice, dtype: float64

In [29]:
df.housestyle = df['housestyle'].astype(int).astype(str)
df.housestyle = df.housestyle.apply(str.strip)

for i in range (0, len(df.housestyle)):
        if df.housestyle[i] == '1.5Fin':
            df.housestyle[i] = 0
        elif df.housestyle[i] == '1.5Unf':
            df.housestyle[i] = 1
        elif df.housestyle[i] == '1Story':
            df.housestyle[i] = 2
        elif df.housestyle[i] == '2.5Fin':
            df.housestyle[i] = 3
        elif df.housestyle[i] == '2.5Unf':
            df.housestyle[i] = 4
        elif df.housestyle[i] == '2Story':
            df.housestyle[i] = 5
        elif df.housestyle[i] == 'SFoyer':
            df.housestyle[i] = 6
        elif df.housestyle[i] == 'SLv1':
            df.housestyle[i] = 7
        else:
            df.housestyle[i] = 8
            
df["is_street_paved"] = pd.get_dummies(df.street, drop_first=True)
df.housestyle = df['housestyle'].astype(str).astype(int)

1460

In [48]:
X = df[['housestyle', 'is_street_paved', 'overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf']]
Y = df['saleprice']

X = sm.add_constant(X, has_constant='add')
results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     771.8
Date:                Sun, 23 Feb 2020   Prob (F-statistic):               0.00
Time:                        21:53:59   Log-Likelihood:                -17499.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1453   BIC:                         3.505e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -1669.1759    251.245     -6.644      0.000   -2162.018   -1176.334
housestyle      -1.335e+04   2009.960     -6.644      0.000   -1.73e+04   -9410.674
is_street_paved  9649.4588    1.6e+04      0.602      0.547   -2.18e+04    4.11e+04
overallqual      2.359e+04   1075.481     21.933      0.000    2.15e+04    2.57e+04
grlivarea          45.2961      2.491     18.185      0.000      40.410      50.182
garagecars       1.451e+04   3023.880      4.798      0.000    8578.399    2.04e+04
garagearea         17.5856     10.498      1.675      0.094      -3.006      38.178
totalbsmtsf        31.5175      2.905     10.851      0.000      25.820      37.215
==============================================================================
Omnibus:                      438.229   Durbin-Watson:                   1.972
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            40766.171
Skew:                          -0.337   Prob(JB):                         0.00
Kurtosis:                      28.878   Cond. No.                     1.01e+18
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 5.7e-27. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

is street paved is not statistically significant. 

In [49]:
X = df[['housestyle', 'overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf']]
Y = df['saleprice']

X = sm.add_constant(X, has_constant='add')
results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     926.5
Date:                Sun, 23 Feb 2020   Prob (F-statistic):               0.00
Time:                        21:54:21   Log-Likelihood:                -17499.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1454   BIC:                         3.504e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1524.1854     71.361    -21.359      0.000   -1664.166   -1384.204
housestyle  -1.219e+04    570.886    -21.359      0.000   -1.33e+04   -1.11e+04
overallqual  2.364e+04   1072.532     22.037      0.000    2.15e+04    2.57e+04
grlivarea      45.3458      2.489     18.218      0.000      40.463      50.228
garagecars   1.454e+04   3022.681      4.812      0.000    8615.034    2.05e+04
garagearea     17.1334     10.468      1.637      0.102      -3.401      37.668
totalbsmtsf    31.5015      2.904     10.848      0.000      25.805      37.198
==============================================================================
Omnibus:                      438.146   Durbin-Watson:                   1.971
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            40648.199
Skew:                          -0.338   Prob(JB):                         0.00
Kurtosis:                      28.841   Cond. No.                     8.09e+17
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 8.92e-27. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

the biggest change was in the constant. the other coefficients are relatively similar. the garage area coefficient is borderline statistically significant/insignificant (.102). the things that influence the sale price most are the overall quality of the house and the amount of cars that the garage holds. this makes sense as overall quality of a house will of course have a positive efffect on the sale price. if a house is in poor condition, then the amount they can ask for is significantly less than it could be because of repairs etc. the garage cars also makes sense as generally the more cars a garage can hold, the larger the house is.